File: langs/go/interpolation.go
package dotlyte

import (
	"fmt"
	"os"
	"regexp"
	"strings"
)

var interpolationPattern = regexp.MustCompile(`\$\{([^}]+)\}`)

// Interpolate resolves ${VAR}, ${VAR:-default}, and ${VAR:?error} references
// within a flat string map. Resolution order: same-file values → context → os.Environ.
// Circular references are detected and cause an InterpolationError.
func Interpolate(data map[string]string, context map[string]string) (map[string]string, error) {
	result := make(map[string]string, len(data))
	resolving := make(map[string]bool)
	resolved := make(map[string]string)

	var resolve func(key string) (string, error)
	resolve = func(key string) (string, error) {
		if val, ok := resolved[key]; ok {
			return val, nil
		}
		if resolving[key] {
			return "", &InterpolationError{
				DotlyteError: DotlyteError{
					Message: fmt.Sprintf("circular reference detected: %s", key),
					Code:    "INTERPOLATION_CIRCULAR",
				},
				Variable: key,
			}
		}

		raw, exists := data[key]
		if !exists {
			// Try context then env
			if val, ok := context[key]; ok {
				return val, nil
			}
			return os.Getenv(strings.ToUpper(key)), nil
		}

		resolving[key] = true
		val, err := resolveString(raw, data, context, resolve)
		if err != nil {
			return "", err
		}
		delete(resolving, key)
		resolved[key] = val
		return val, nil
	}

	for key := range data {
		val, err := resolve(key)
		if err != nil {
			return nil, err
		}
		result[key] = val
	}

	return result, nil
}

func resolveString(
	s string,
	data map[string]string,
	context map[string]string,
	resolve func(string) (string, error),
) (string, error) {
	// Handle escaped $$
	s = strings.ReplaceAll(s, "$$", "\x00DOLLAR\x00")

	var lastErr error
	result := interpolationPattern.ReplaceAllStringFunc(s, func(match string) string {
		if lastErr != nil {
			return match
		}

		inner := match[2 : len(match)-1] // strip ${ and }

		var varName, fallback, errMsg string
		var hasFallback, hasError bool

		if idx := strings.Index(inner, ":?"); idx != -1 {
			varName = inner[:idx]
			errMsg = inner[idx+2:]
			hasError = true
		} else if idx := strings.Index(inner, ":-"); idx != -1 {
			varName = inner[:idx]
			fallback = inner[idx+2:]
			hasFallback = true
		} else {
			varName = inner
		}

		varName = strings.TrimSpace(varName)

		// Try resolving from same-file data
		if _, exists := data[strings.ToLower(varName)]; exists {
			val, err := resolve(strings.ToLower(varName))
			if err != nil {
				lastErr = err
				return match
			}
			if val != "" {
				return val
			}
		}

		// Try context
		if val, ok := context[strings.ToLower(varName)]; ok && val != "" {
			return val
		}

		// Try environment
		if val := os.Getenv(varName); val != "" {
			return val
		}
		// Also try uppercase
		if val := os.Getenv(strings.ToUpper(varName)); val != "" {
			return val
		}

		if hasError {
			lastErr = &InterpolationError{
				DotlyteError: DotlyteError{
					Message: fmt.Sprintf("required variable '%s': %s", varName, errMsg),
					Code:    "INTERPOLATION_REQUIRED",
				},
				Variable: varName,
			}
			return match
		}

		if hasFallback {
			return fallback
		}

		return ""
	})

	if lastErr != nil {
		return "", lastErr
	}

	result = strings.ReplaceAll(result, "\x00DOLLAR\x00", "$")
	return result, nil
}

File: langs/rust/src/interpolation.rs
//! Variable interpolation engine for DOTLYTE v2.
//!
//! Supports `${VAR}`, `${VAR:-default}`, `${VAR:?error}`, and `$$` escape.

use std::collections::{HashMap, HashSet};
use std::env;

use crate::errors::DotlyteError;

/// Interpolate `${VAR}` references in a flat string map.
///
/// Resolution order: same-file values → context → `std::env::var`.
/// Circular references produce an `InterpolationError`.
pub fn interpolate(
    data: &HashMap<String, String>,
    context: &HashMap<String, String>,
) -> crate::Result<HashMap<String, String>> {
    let mut resolved: HashMap<String, String> = HashMap::new();
    let mut resolving: HashSet<String> = HashSet::new();

    fn resolve(
        key: &str,
        data: &HashMap<String, String>,
        context: &HashMap<String, String>,
        resolved: &mut HashMap<String, String>,
        resolving: &mut HashSet<String>,
    ) -> crate::Result<String> {
        if let Some(val) = resolved.get(key) {
            return Ok(val.clone());
        }
        if resolving.contains(key) {
            return Err(DotlyteError::InterpolationError {
                variable: key.to_string(),
                message: format!("circular reference detected: {key}"),
            });
        }

        let raw = match data.get(key) {
            Some(r) => r.clone(),
            None => {
                if let Some(v) = context.get(key) {
                    return Ok(v.clone());
                }
                return Ok(env::var(key.to_uppercase()).unwrap_or_default());
            }
        };

        resolving.insert(key.to_string());
        let val = resolve_string(&raw, data, context, resolved, resolving)?;
        resolving.remove(key);
        resolved.insert(key.to_string(), val.clone());
        Ok(val)
    }

    let keys: Vec<String> = data.keys().cloned().collect();
    for key in &keys {
        resolve(key, data, context, &mut resolved, &mut resolving)?;
    }

    Ok(resolved)
}

fn resolve_string(
    s: &str,
    data: &HashMap<String, String>,
    context: &HashMap<String, String>,
    resolved: &mut HashMap<String, String>,
    resolving: &mut HashSet<String>,
) -> crate::Result<String> {
    let s = s.replace("$$", "\x00DOLLAR\x00");
    let mut result = String::with_capacity(s.len());
    let mut chars = s.chars().peekable();

    while let Some(ch) = chars.next() {
        if ch == '$' && chars.peek() == Some(&'{') {
            chars.next(); // consume '{'
            let mut inner = String::new();
            let mut depth = 1;
            for ch in chars.by_ref() {
                if ch == '{' {
                    depth += 1;
                } else if ch == '}' {
                    depth -= 1;
                    if depth == 0 {
                        break;
                    }
                }
                inner.push(ch);
            }

            let replacement = resolve_reference(&inner, data, context, resolved, resolving)?;
            result.push_str(&replacement);
        } else {
            result.push(ch);
        }
    }

    Ok(result.replace("\x00DOLLAR\x00", "$"))
}

fn resolve_reference(
    inner: &str,
    data: &HashMap<String, String>,
    context: &HashMap<String, String>,
    resolved: &mut HashMap<String, String>,
    resolving: &mut HashSet<String>,
) -> crate::Result<String> {
    let (var_name, fallback, err_msg) = if let Some(idx) = inner.find(":?") {
        (&inner[..idx], None, Some(&inner[idx + 2..]))
    } else if let Some(idx) = inner.find(":-") {
        (&inner[..idx], Some(&inner[idx + 2..]), None)
    } else {
        (inner, None, None)
    };

    let var_name = var_name.trim();
    let lower = var_name.to_lowercase();

    // Try same-file
    if data.contains_key(&lower) {
        let val = resolve(var_name, data, context, resolved, resolving)?;
        if !val.is_empty() {
            return Ok(val);
        }
    }

    // Try context
    if let Some(v) = context.get(&lower) {
        if !v.is_empty() {
            return Ok(v.clone());
        }
    }

    // Try env
    if let Ok(v) = env::var(var_name) {
        if !v.is_empty() {
            return Ok(v);
        }
    }
    if let Ok(v) = env::var(var_name.to_uppercase()) {
        if !v.is_empty() {
            return Ok(v);
        }
    }

    // Not found
    if let Some(msg) = err_msg {
        return Err(DotlyteError::InterpolationError {
            variable: var_name.to_string(),
            message: format!("required variable '{var_name}': {msg}"),
        });
    }

    if let Some(fb) = fallback {
        return Ok(fb.to_string());
    }

    Ok(String::new())
}

fn resolve(
    key: &str,
    data: &HashMap<String, String>,
    context: &HashMap<String, String>,
    resolved: &mut HashMap<String, String>,
    resolving: &mut HashSet<String>,
) -> crate::Result<String> {
    let lower = key.to_lowercase();
    if let Some(val) = resolved.get(&lower) {
        return Ok(val.clone());
    }
    if resolving.contains(&lower) {
        return Err(DotlyteError::InterpolationError {
            variable: key.to_string(),
            message: format!("circular reference detected: {key}"),
        });
    }
    let raw = match data.get(&lower) {
        Some(r) => r.clone(),
        None => {
            if let Some(v) = context.get(&lower) {
                return Ok(v.clone());
            }
            return Ok(env::var(key.to_uppercase()).unwrap_or_default());
        }
    };
    resolving.insert(lower.clone());
    let val = resolve_string(&raw, data, context, resolved, resolving)?;
    resolving.remove(&lower);
    resolved.insert(lower, val.clone());
    Ok(val)
}
